src/FlowMeter.cpp

```diff
@@ -58,16 +63,29 @@ void FlowMeter::count() {
     this->_currentPulses++;                                                 //!< this should be called from an interrupt service routine
 }
 
+void FlowMeter::pause() {
+    detachInterrupt(_pin);
+    this->_currentPulses = 0;
+}
+
+void FlowMeter::resume() {
+    attachInterrupt(_pin, _interruptCallback, RISING);
+}
+
 void FlowMeter::reset() {
-    cli();                                                                  //!< going to change interrupt variable(s)
+    detachInterrupt(_pin);                                                  //!< going to change interrupt variable(s)
```

Comment:
This is semantically different, so it doesn't do what's required here! 

Detaching this particular interrupt handler will do nothing for other interrupts that the application might have registered. 

The intention here is to stop (and subsequently restart) interrupt handling *altogether*, not just for this particular source/handler. 

In short: other interrupts might still fire, interrupting FlowMeter's pulse (i.e. missing pulses inbetween counting steps) counting. 

If this works reliably for you, it's probably because your application doesn't use other interrupt handlers or doesn't have interrupts firing at the same time. 
OTOH, you might be losing pulses all along. 

What we need at this point is a portable way to stop and start interrupt handling, preferably for all sorts of platforms/microcontrollers. 

Please correct me if I'm wrong in this regard!

Reply:
There is a good chance I was missing pulses because there was a 5-10% error margin sometimes. This definitely could've been the reason as to why that was happening. When I have some time next, I'll explore to see if there's a portable way to do what `cli` was doing then instead!

Reply:
From what I understand, when you have an ISR interrupt, it pauses all other interrupts implicitly. Reading up on the topic here: https://arduino.stackexchange.com/questions/30968/how-do-interrupts-work-on-the-arduino-uno-and-similar-boards

I think in that regard, when we enter the interrupt code, the other interrupts should be paused implicitly.

Reply:
The stack exchange answer above says:

> if multi-byte fields are being updated by an ISR then you may need to disable interrupts

That's also how I was introduced to interrupt handling.
Whenever a handler routine requires more than one cpu cycle, it *might* be interrupted in the middle of it's operations. 

We're only updating a single word here, but different platforms might still handle that differently. 

Interrupts are queued, handling them will continue after re-enabling. 
At least that's how the usual microcontrollers do it.

This way, no interrupt is lost while at the same time no handler is being interrupted – provided the handler routines are kept short.

Reply:
And yes, `cli()` clears all interrupts (by clearing the mask), but doesn't interrupt the handler references.

This way, no interrupts (except *reset*) are handled – but all of them are queued (only limited by queue length).